### PR TITLE
appflowy: init at 0.0.2

### PR DESCRIPTION
**Pull request by DarkOnion0**
###### Motivation for this change
Add AppFlowy to nixpkgs according to this packaging request #146724

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/office/appflowy/default.nix [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  installPhase = ''
+    mkdir -p $out/opt/
+    mkdir -p $out/share/applications/
+    mkdir -p $out/bin/
+
+    # Copy archive contents to the outpout directory
+    cp -r ./* $out/opt/
+
+    # Create the desktop file
+    echo -e "[Desktop Entry] \nName=AppFlowy \nComment=An Open Source Alternative to Notion \nIcon=$out/opt/flowy_logo.svg \nExec=$out/bin/appflowy \nCategories=Office \nType=Application \nTerminal=false" > $out/share/applications/appflowy.desktop
```
Comment (legendofmiracles) [inline pkgs/applications/office/appflowy/default.nix:34]: Use the desktopItems attribute and `copyDesktopItems`. See [here](https://github.com/NixOS/nixpkgs/blob/75a86ab9f5691930f79de6c72e9f80eb280cf2dd/pkgs/games/dhewm3/default.nix#L27) and [here](https://github.com/NixOS/nixpkgs/blob/75a86ab9f5691930f79de6c72e9f80eb280cf2dd/pkgs/games/dhewm3/default.nix#L24) for an example

File: pkgs/applications/office/appflowy/default.nix [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    # Add missing libraries to appflowy using the ones it comes with
+    makeWrapper $out/opt/app_flowy $out/opt/app_flowy_warpped --set LD_LIBRARY_PATH "$out/opt/lib/"
+
+    # Symlink wrapper
+    ln -s $out/opt/app_flowy_warpped $out/bin/appflowy
```
Comment (legendofmiracles) [inline pkgs/applications/office/appflowy/default.nix:40]:
```suggestion
    # Add missing libraries to appflowy using the ones it comes with
    makeWrapper $out/opt/app_flowy $out/bin/appflowy \
      --set LD_LIBRARY_PATH "$out/opt/lib/"
```
Why do we need to link it? Can't we just immediately write the wrapper to the location?

Comment (DarkOnion0) [reply pkgs/applications/office/appflowy/default.nix:40]: Why not ? I will try this

File: pkgs/applications/office/appflowy/default.nix [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+  buildInputs = [
+      gtk3
+      openssl
+  ];
+
+  dontBuild = true;
+  dontConfigure = true;
+
+  installPhase = ''
```
Comment (AndersonTorres) [inline pkgs/applications/office/appflowy/default.nix:25]: Don't forget the hooks.
```suggestion
  installPhase = ''
    runHook preInstall
    
```

File: pkgs/applications/office/appflowy/default.nix [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    # Create the desktop file
+    echo -e "[Desktop Entry] \nName=AppFlowy \nComment=An Open Source Alternative to Notion \nIcon=$out/opt/flowy_logo.svg \nExec=$out/bin/appflowy \nCategories=Office \nType=Application \nTerminal=false" > $out/share/applications/appflowy.desktop
+
+    # Add missing libraries to appflowy using the ones it comes with
+    makeWrapper $out/opt/app_flowy $out/opt/app_flowy_warpped --set LD_LIBRARY_PATH "$out/opt/lib/"
+
+    # Symlink wrapper
+    ln -s $out/opt/app_flowy_warpped $out/bin/appflowy
+  '';
```
Comment (AndersonTorres) [inline pkgs/applications/office/appflowy/default.nix:41]:
```suggestion
  
  runHook postInstall
  '';
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
There is a source release of this?

File: pkgs/applications/office/appflowy/default.nix (hunk 1 above)
Comment (DarkOnion0) [reply pkgs/applications/office/appflowy/default.nix]: I applied your change, but the `.desktop` file doesn't seem to work when I try to launch it through my file explorer. 
Is it normal ?

**Comment by DarkOnion0** [comment]
> There is a source release of this?

Yes, there is one, but I haven't found a way to package it :disappointed:

File: pkgs/applications/office/appflowy/default.nix (hunk 1 above)
Comment (legendofmiracles) [reply pkgs/applications/office/appflowy/default.nix]: It seems to start it for me, be sure to really install everything.

Comment (DarkOnion0) [reply pkgs/applications/office/appflowy/default.nix]: Okay thanks :sweat_smile:

**Comment by legendofmiracles** [comment]
When I run it, it's only a black screen, with this error:
```
[ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: ProcessException: Failed to find "xdg-user-dir" in the search path.
  Command: xdg-user-dir
```
After I install xdg-user-dirs (so please add that dependency to the wrapper 😄) and create the ~/Documents dir it runs fine though

**Comment by DarkOnion0** [comment]
> When I run it, it's only a black screen, with this error:
> 
> ```
> [ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: ProcessException: Failed to find "xdg-user-dir" in the search path.
>   Command: xdg-user-dir
> ```
> 
> After I install xdg-user-dirs (so please add that dependency to the wrapper :smile:) and create the ~/Documents dir it runs fine though

It is normally fixed, at least I hope, because I never had this issue before

File: pkgs/applications/office/appflowy/default.nix [hunk 5]
```diff
@@ -1,4 +1,4 @@
-{ stdenv, lib, fetchzip, autoPatchelfHook, makeWrapper, gtk3, openssl, copyDesktopItems, makeDesktopItem }:
+{ stdenv, lib, fetchzip, autoPatchelfHook, makeWrapper, gtk3, openssl, copyDesktopItems, makeDesktopItem, xdg-user-dirs }:
```
Comment (AndersonTorres) [inline pkgs/applications/office/appflowy/default.nix:1]: I suggest put each attribute in a single line here. It helps in future diffs.
```suggestion
{ lib,
  stdenv,
  fetchzip,
  autoPatchelfHook,
  makeWrapper,
  gtk3,
  openssl,
  copyDesktopItems,
  makeDesktopItem,
  xdg-user-dirs 
}:
```

File: pkgs/applications/office/appflowy/default.nix [hunk 6]
```diff
@@ -35,11 +35,17 @@ stdenv.mkDerivation rec {
 
     # Create the desktop file
```
Comment (AndersonTorres) [inline pkgs/applications/office/appflowy/default.nix:36]: Useless comment(?)

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/applications/office/appflowy/default.nix (hunk 6 above)
Comment (DarkOnion0) [reply pkgs/applications/office/appflowy/default.nix:36]: You're right, sorry :sweat_smile:

File: pkgs/applications/office/appflowy/default.nix [hunk 7]
```diff
@@ -0,0 +1,78 @@
+{ stdenv,
+  lib,
+  fetchzip,
+  autoPatchelfHook,
+  makeWrapper,
+  copyDesktopItems,
+  makeDesktopItem,
+
+  # Dynamic libs
```
Comment (AndersonTorres) [inline pkgs/applications/office/appflowy/default.nix:9]: Unneeded, but OK.

**Review by AndersonTorres — APPROVED** [review]

**Comment by AndersonTorres** [comment]
Next thing to be done is to rebase the commits into one.

File: pkgs/applications/office/appflowy/default.nix (hunk 7 above)
Comment (AndersonTorres) [inline pkgs/applications/office/appflowy/default.nix:9]:
```suggestion
```

Unneeded comment

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Now squash all commits into one.

**Comment by AndersonTorres** [comment]
Now rebase all commits into only one.

**Comment by DarkOnion0** [comment]
> Now rebase all commits into only one.

Hey, I just merge all the commits message in only one, I think this should be okay now. I apologize for the delay, but I had some trouble with neovim and git :sweat_smile:

**Comment by DarkOnion0** [comment]
Hi @AndersonTorres is there any problems or things to change now since ofborg has successfully finished his checks ?

**Comment by AndersonTorres** [comment]
Nope. All OK!

**Pull request merged**